### PR TITLE
project crud functionality

### PR DESCRIPTION
**Pull request by operatorjen**

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+const ALPHANUM_MATCH = /[^a-z0-9]+/gi;
+
+// Various utility functions
+
+// Generic object extension
+Object.defineProperty(Object.prototype, 'extend', {
```
Comment (seanmonstar) [inline lib/utils.js]: omgwat!

File: test/test.projects.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        session: {
+          email: 'test@test.org'
+        },
+        body: {
+          title: 'My Project'
+        }
+      };
+
+      projects.get(req, db, 1, function(err, foundProject) {
+        foundProject.should.not.be.null;
```
Comment (seanmonstar) [inline test/test.projects.js]: if it were null, wouldn't this throw an error trying to access `should`?

Comment (operatorjen) [inline test/test.projects.js]: yes

File: test/test.projects.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          title: 'My Project2'
+        }
+      };
+
+      projects.update(req, db, 1, function(err, project) {
+        project.title.should.equal(req.body.title);
+      });
+    });
+
+    it('does not update specific project', function() {
```
Comment (seanmonstar) [inline test/test.projects.js]: why doesn't it update? the email is different?

Comment (operatorjen) [inline test/test.projects.js]: yep projects only updateable by the author for now

File: test/test.projects.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      };
+
+      projects.update(req, db, 1, function(err, project) {
+        project.should.equal(false);
+      });
+    });
+  });
+
+  describe('DELETE /project/:id', function() {
+    it('does not delete a project', function() {
```
Comment (seanmonstar) [inline test/test.projects.js]: same as above. test description doesn't tell me why it shouldn't delete. email is different?

Comment (operatorjen) [inline test/test.projects.js]: i'll add to the 'it' description

File: lib/projects.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    return callback(null, JSON.parse(project));
+  });
+};
+
+/* Project Add
+ * Requires: web request, db connection
+ * Returns: A listing of the user's projects
+ */
+exports.add = function(req, db, callback) {
+  db.incr('project:' + req.session.email, function(err, id) {
```
Comment (seanmonstar) [inline lib/projects.js:63]: no `err` handling here?

File: lib/projects.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    var project = {
+      id: identifier,
+      title: req.body.title,
+      identifier: identifier,
+      author: req.session.email,
+      created: new Date().getTime()
+    };
+
+    db.set('project:' + req.session.email + ':' + identifier, JSON.stringify(project));
```
Comment (seanmonstar) [inline lib/projects.js]: does this throw errors? if so, should it be catch and passed to `callback`?

File: lib/projects.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  var projectKey = 'project:' + req.session.email + ':' + identifier;
+
+  db.get(projectKey, function(err, project) {
+    if (err) {
+      return callback(err);
+    }
+
+    var updatedProject = utils.updateObject(req, project, ALLOWED_FIELDS);
+
+    db.set(projectKey, updatedProject);
```
Comment (seanmonstar) [inline lib/projects.js:108]: does this throw errors? if so, should it be catch and passed to `callback`?

File: lib/projects.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    var project = {
+      id: identifier,
+      title: req.body.title,
+      identifier: identifier,
+      author: req.session.email,
+      created: new Date().getTime()
+    };
+
+    db.set('project:' + req.session.email + ':' + identifier, JSON.stringify(project), function(errAddProj) {
+      if (errAddProj) {
```
Comment (seanmonstar) [inline lib/projects.js:79]: indentically challenged

Comment (operatorjen) [inline lib/projects.js:79]: OMG SEMICOLONS.

**Pull request merged**